src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using System.Web.Hosting;
+using System.Web.Http;
+using Microsoft.Azure.WebJobs.Host.Timers;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    public class WebScriptHostExceptionHandler : IWebJobsExceptionHandler
+    {
+        public void Initialize(JobHost host)
+        {
+        }
+
+        public Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod)
+        {
+            WebScriptHostManager manager = GetManager();
+            manager?.Instance.TraceWriter.Error("A function timeout has occurred. Host is shutting down.");
+            manager?.Instance.TraceWriter.Flush();
+
+            // Give the manager some time to shut down gracefully.
+            Task.Delay(timeoutGracePeriod).ContinueWith((ct) =>
+            {
+                HostingEnvironment.InitiateShutdown();
+            });
+
+            manager?.Stop();
```

Comment:
Should this be called before initiating the delay and shutdown? Seems you'd want to stop taking on new work immediately